src/Domain/CoordinateMaps/CubicScale.cpp

```diff
@@ -101,168 +168,258 @@ boost::optional<std::array<tt::remove_cvref_wrap_t<T>, 1>> CubicScale::inverse(
   // version, here we utilize the boost implementation, as it includes
   // additional checks for zero derivative, checks on bounds, and can implement
   // bisection if necessary.
-  return {
-      {{outer_boundary_ * RootFinder::newton_raphson(
-                              cubic_and_deriv, initial_guess, 0.0, 1.0, 14)}}};
+  const double scale_factor =
+      RootFinder::newton_raphson(cubic_and_deriv, initial_guess, 0.0, 1.0, 14) /
+      target_dimensionless_radius;
+
+  std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+  for (size_t i = 0; i < Dim; ++i) {
+    gsl::at(result, i) = scale_factor * gsl::at(target_coords, i);
+  }
+  return {std::move(result)};
 }
 
+template <size_t Dim>
 template <typename T>
-std::array<tt::remove_cvref_wrap_t<T>, 1> CubicScale::frame_velocity(
-    const std::array<T, 1>& source_coords, const double time,
+std::array<tt::remove_cvref_wrap_t<T>, Dim> CubicScale<Dim>::frame_velocity(
+    const std::array<T, Dim>& source_coords, const double time,
     const std::unordered_map<
         std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
         map_list) const noexcept {
   const auto dt_a_of_t = map_list.at(f_of_t_a_)->func_and_deriv(time)[1][0];
+
+  if (functions_of_time_equal_) {
+    // optimization for linear radial scaling
+    std::array<tt::remove_cvref_wrap_t<T>, Dim> result{};
+    for (size_t i = 0; i < Dim; ++i) {
+      gsl::at(result, i) = dt_a_of_t * gsl::at(source_coords, i);
+    }
+    return result;
+  }
+
   const auto dt_b_of_t = map_list.at(f_of_t_b_)->func_and_deriv(time)[1][0];
-  const auto frame_vel =
-      source_coords[0] *
-      (dt_a_of_t +
-       (dt_b_of_t - dt_a_of_t) * square(source_coords[0] / outer_boundary_));
 
-  return {{frame_vel}};
+  tt::remove_cvref_wrap_t<T> r_squared =
+      square(dereference_wrapper(source_coords[0]));
+  for (size_t i = 1; i < Dim; ++i) {
+    r_squared += square(dereference_wrapper(gsl::at(source_coords, i)));
+  }
+  // Reuse r^2 allocation
+  r_squared =
```

Comment:
I realize you're reusing for optimization, but I'm confused about the name. Is `r_squared` ever equal to source coord radius square? If so, shouldn't it be `rho_squared`? `r^2` makes it sound like it's the square of the result, using the notation of the help text, but then a sqrt is never taken, so it can't mean that

Reply:
I changed it to `rho_squared` while addressing Jordan's comments, which was pushed while you reviewing